server/etcdserver/api/v3rpc/watch.go

```diff
@@ -267,13 +268,27 @@ func (sws *serverWatchStream) recvLoop() error {
 				creq.RangeEnd = []byte{}
 			}
 
-			if !sws.isWatchPermitted(creq) {
+			err := sws.isWatchPermitted(creq)
+			if err != nil {
+				var cancelReason string
+				switch err {
+				case auth.ErrInvalidAuthToken:
+					cancelReason = rpctypes.ErrGRPCInvalidAuthToken.Error()
+				case auth.ErrAuthOldRevision:
+					cancelReason = rpctypes.ErrGRPCAuthOldRevision.Error()
+				default:
+					if err != auth.ErrPermissionDenied {
+						sws.lg.Panic("unexpected error code", zap.Error(err))
```

Comment:
It seems that it's also possible to get an [ErrUserEmpty](https://github.com/etcd-io/etcd/blob/1222269ebaf7ac7f9e8dca79bf1582bba1f0de74/server/auth/store.go#L849) response.

Reply:
It might not be good to panic etcd server.  I suggest to change it to an error log instead, and return `ErrPermissionDenied` to clients.

Reply:
Good catch, thanks! Let me fix.